File: retroauth-android/src/test/java/com/andretietz/retroauth/AuthAccountManagerTest.java
package com.andretietz.retroauth;

import android.accounts.Account;
import android.accounts.AccountManager;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.annotation.Config;

import static com.andretietz.retroauth.testhelper.Helper.setMember;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 23)
public class AuthAccountManagerTest {


    private AuthAccountManager authAccountManager;
    private AccountManager accountManager;

    @Before
    public void setup() throws NoSuchFieldException, IllegalAccessException {
        ContextManager.get(RuntimeEnvironment.application);
        accountManager = mock(AccountManager.class);
        authAccountManager = new AuthAccountManager();
        setMember(authAccountManager, "accountManager", accountManager);
    }


    @SuppressWarnings("MissingPermission")
    @Test
    public void accountAmount() throws NoSuchFieldException, IllegalAccessException {
        AccountManager accountManager = mock(AccountManager.class);
        setMember(this.authAccountManager, "accountManager", accountManager);

        when(accountManager.getAccountsByType(anyString()))
                .thenReturn(new Account[]{mock(Account.class)});

        assertEquals(this.authAccountManager.accountAmount("asd"), 1);
    }

    @Test
    public void getActiveAccountFailing() {
        authAccountManager.resetActiveAccount("accountType");
        Account account = authAccountManager.getActiveAccount("accountType");
        assertNull(account);
    }

    @SuppressWarnings("MissingPermission")
    @Test
    public void getActiveAccountSuccess() {
        when(accountManager.getAccountsByType(anyString()))
                .thenReturn(new Account[]{new Account("accountName", "accountType")});
        Account activeAccount = authAccountManager
                .setActiveAccount("accountType", "accountName");

        Account account = authAccountManager.getActiveAccount("accountType");


        assertNotNull(account);
        assertEquals("accountType", account.type);
        assertEquals("accountName", account.name);
        assertEquals(activeAccount, account);
    }

    @SuppressWarnings("MissingPermission")
    @Test
    public void getAccountByNameSuccess() {
        Account testAccount = new Account("accountName", "accountType");
        when(accountManager.getAccountsByType(anyString()))
                .thenReturn(new Account[]{testAccount});
        Account account = authAccountManager.getAccountByName("accountType", "accountName");

        assertNotNull(account);
        assertEquals(testAccount, account);
    }

    @SuppressWarnings("MissingPermission")
    @Test
    public void getAccountByNameFail() {
        Account testAccount = new Account("accountName", "accountType");
        when(accountManager.getAccountsByType(anyString()))
                .thenReturn(new Account[]{});
        Account account = authAccountManager.getAccountByName("accountType", "accountName");

        assertNull(account);
    }

    @SuppressWarnings("MissingPermission")
    @Test
    public void getActiveUserData() {
        Account testAccount = new Account("accountName", "accountType");
        when(accountManager.getAccountsByType(anyString()))
                .thenReturn(new Account[]{testAccount});
        when(accountManager.getUserData(any(Account.class), anyString()))
                .thenReturn("value");
        authAccountManager
                .setActiveAccount("accountType", "accountName");

        String data = authAccountManager.getActiveUserData("accountType", "key");
        assertNotNull(data);
        assertEquals("value", data);
    }
}

File: retroauth-android/src/test/java/com/andretietz/retroauth/AuthAccountManagerTest.kt
package com.andretietz.retroauth

import android.accounts.Account
import android.accounts.AccountManager
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class AuthAccountManagerTest {

    private val accountManager: AccountManager = mock(AccountManager::class.java)
    private val authAccountManager: AuthAccountManager = AuthAccountManager(RuntimeEnvironment.application, accountManager)


    @Test
    @Throws(NoSuchFieldException::class, IllegalAccessException::class)
    fun accountAmount() {
        `when`(accountManager.getAccountsByType(anyString()))
                .thenReturn(arrayOf(mock(Account::class.java)))

        assertEquals(this.authAccountManager.accountAmount("asd").toLong(), 1)
    }

    @Test
    fun getActiveAccountFailing() {
        authAccountManager.resetActiveAccount("accountType")
        val account = authAccountManager.getActiveAccount("accountType")
        assertNull(account)
    }

    @Test
    fun getActiveAccountSuccess() {
        `when`(accountManager.getAccountsByType(anyString()))
                .thenReturn(arrayOf(Account("accountName", "accountType")))
        val activeAccount = authAccountManager
                .setActiveAccount("accountType", "accountName")

        val account = authAccountManager.getActiveAccount("accountType")


        assertNotNull(account)
        assertEquals("accountType", account!!.type)
        assertEquals("accountName", account.name)
        assertEquals(activeAccount, account)
    }

    @Test
    fun getAccountByNameSuccess() {
        val testAccount = Account("accountName", "accountType")
        `when`(accountManager.getAccountsByType(anyString()))
                .thenReturn(arrayOf(testAccount))
        val account = authAccountManager.getAccountByName("accountType", "accountName")

        assertNotNull(account)
        assertEquals(testAccount, account)
    }

    @Test
    fun getAccountByNameFail() {
//        val testAccount = Account("accountName", "accountType")
        `when`(accountManager.getAccountsByType(anyString())).thenReturn(arrayOf())
        val account = authAccountManager.getAccountByName("accountType", "accountName")

        assertNull(account)
    }

    @Test
    fun getActiveUserData() {
        val testAccount = Account("accountName", "accountType")
        `when`(accountManager.getAccountsByType(anyString())).thenReturn(arrayOf(testAccount))
        `when`(accountManager.getUserData(any(Account::class.java), anyString())).thenReturn("value")
        authAccountManager.setActiveAccount("accountType", "accountName")

        val data = authAccountManager.getActiveUserData("accountType", "key")
        assertNotNull(data)
        assertEquals("value", data)
    }
}
